### DISCUSSION TITLE
Adaptive Bitrate Streaming (ABS/ABR) in RustDesk

### DISCUSSION THREAD

**Discussion by Heap-Hop** [rustdesk/rustdesk#792] [General]
We can control bitrate and quantizer manually in client mode, but we haven't implemented ABR yet.

There are some stable ABS algorithms in HTTP streaming media, like MPEG-DASH/HLS, or like TFRC/Kalman-filter in WebRTC.

We are considering how to port a suitable ABS/ABR algorithm to RustDesk.

Some informations of RustDesk for now:
- Sending data over TCP (QUIC and UDP may be used in the future).
- Encoder can receive bitrate dynamically (we use VP9 current, and H264/H265 under development).
- In multi clients mode, one video server will send the same frame multi times to each client, the same bitrate will be shared around clients.

You're welcome to share your ideas.

and thx @WangHoi #280 #79 @AndrewMD5

**Comment by rustdesk** [comment]
RTP used in moonlight: https://github.com/moonlight-stream/moonlight-common-c/tree/e62dc56047b038e5f2a5404b023fec453bf1bf8a/src

https://github.com/SunshineStream/Sunshine

**Comment by BobAnkh** [comment]
Interesting to use ABR in remote desk. I have been researching on ABR video streaming and have to admit that kalman filter used in GCC in webrtc is not as efficient as the paper describes(but might be a easy one to be implemented). DASH is easy, but it is not designed for Real-Time Communication I am afraid. 

Therefor I refer you to implement those CC and bit rate control for RTC(at this point GCC might be an easy one to implement lol, but *NADA*(PV'13)might also be a good choice). As you have been using VP9, I highly recommend you to the paper *Salsify*(NSDI'18, top conference in network) which combines video codec with transport potocal to provide low latency. Besides, SVC should be a good choice for you to consider or start with, which is easy and efficient. And there is a evolutional version of SVC as layerd codec in this year's top conference *Swift*(NSDI'22).

BTW, I wonder what is the typical frame rate of rustdesk?

**Reply by Heap-Hop** [reply]
Many thanks!

> typical frame rate of rustdesk

Currently, 30 frames for dynamic desktop.

**Reply by BobAnkh** [reply]
> Many thanks!
> 
> > typical frame rate of rustdesk
> 
> Currently, 30 frames for dynamic desktop.

Ask for the frame rate just to check if there is some room for optimize. I had been working on ultra-low latency and high resolution interactive video streaming, so there is actually many dimensions to optimize the QoE of a RTC application.

**Reply by Heap-Hop** [reply]
Thanks again, we do need advice on ABR.

Here is our plan (just refer to some implementations of webrtc)

1. Use UDP/RTP instead of current TCP, in video frame transmission.

2. Using RTP header to implement ABR.

3. Add some ABR algorithms.

If you have better and more advanced ideas please share with us, and welcome to contribute.  :)

**Reply by BobAnkh** [reply]
Thanks for your reply. 

UDP/RTP both are good choice. UDP might be easirer and little loss on downlink might not always be handled on RTC app(actually there is a method called skip-frame to skip some frame to the next I-frame on purpose in order to deal with the stall, and it might be easier for you to integrate while you are using functinal codec) and user-space stack is much more easier to be constructed, controlled and *updated* IMO.

For those simple ABR, there might be no need on packet-header adjustment. Most of the simple ABR algo uses statistics collected(acks) on server-side(i.e. graph-sending side).

Again, SVC might be considered as an app-level approach and it is only a codec method so that means easy to implement. Kalman-filter on webrtc is actually as a part of the underlying GCC. (don't know if you have implemented any congestion control algorithms, which might be **very important** on RTC apps). I haven't used RTP-header to do ABR so can't tell whether this is a good choice. webrtc does this is because   it is a 4-7 layerd stack. What information you want to be carried in RTP-header?

Glad to help you and be an advisor on whatever related to video streaming and congestion control or real-time communication(have been researching on these fields for years). I'm not familiar with the project(it's awesome but a bit complicated, though), so might be hard to contribute code.

Feel free to discuss with me or even have a meeting with me 😄

**Reply by Heap-Hop** [reply]
Cool! 
We will follow more about SVC.

> Using RTP header to implement ABR.

My mistake about 'RTP header', it should be :
```
1. Use UDP/RTP instead of current TCP, in video frame transmission.

2. Using RTCP to implement ABR.

3. Add some ABR algorithms.
```
---

> congestion control algorithms

Not yet.

Everything about ABR is still in the research stage.

We have implemented a workaround 'ABR' in current TCP channel. https://github.com/rustdesk/rustdesk/pull/768/commits/a7769da317da35aac593ffe36f59d10ac935d068
But it is far away from real ARB/congestion control.

In short term, seems like we should implement UDP first.

**Comment by jonatino** [comment]
@rustdesk SRT looks very promising and fits the criteria. 

https://www.haivision.com/products/srt-secure-reliable-transport/

**Reply by jonatino** [reply]
RTMP vs SRT both TCP: https://www.youtube.com/watch?v=LGLkIXqBPe8

**Comment by evaneus** [comment]
Have you guys considered looking at/contributing to [webrtc-rs](https://github.com/webrtc-rs/webrtc)?
Work is ongoing, e.g. [rate controller PR](https://github.com/algesten/str0m/pull/47)  but seems like a long term good fit?
